tinytodo/src/policy_store.rs

```diff
@@ -15,84 +15,129 @@
-                            let q = AppQuery::new(kind, send);
-                            tx.blocking_send(q).expect("Failed to send");
-                        }
-                        EventKind::Remove(_) => {
-                            let (send, _recv) = tokio::sync::oneshot::channel();
-                            let kind = AppQueryKind::UpdatePolicySet;
-                            let q = AppQuery::new(kind, send);
-                            tx.blocking_send(q).expect("Failed to send");
-                            let (send, _recv) = tokio::sync::oneshot::channel();
-                            let kind = AppQueryKind::ResetWatch;
-                            let q = AppQuery::new(kind, send);
-                            tx.blocking_send(q).expect("Failed to send");
-                        }
-                        _ => (),
-                    }
-                }
-                Err(err) => error!("Error receiving filesystem event: {}", err),
+pub async fn spwan_watcher(
+    policy_set: impl AsRef<Path>,
+    schema: impl AsRef<Path>,
+    tx: Sender<AppQuery>,
+) {
+    let w = PolicySetWatcher {
+        policy_set: PathBuf::from(policy_set.as_ref()),
+        schema: PathBuf::from(schema.as_ref()),
+        tx,
+    };
+    tokio::spawn(async move { watcher_supervisor(w).await });
+}
+
+// This supervises the watcher task, reporting any errors and respawning the watcher
+async fn watcher_supervisor(w: PolicySetWatcher) {
+    loop {
+        let cloned = w.clone();
+        let handle = tokio::spawn(async { watcher(cloned).await });
+        match handle.await {
+            Ok(f) => match f {
+                Ok(a) => match a {},
+                Err(e) => debug!("Policy Set File Watcher died due to: {e}, respawning..."),
             },
-            config,
-        )
-        .expect("Failed to create watcher");
-
-        let mut s = Self {
-            watcher,
-            path: PathBuf::from(path),
-        };
-        s.set_watch();
-        s
+            Err(e) => error!("Join Error: {e}"),
+        }
     }
+}
 
-    #[tracing::instrument]
-    pub fn set_watch(&mut self) {
-        if let Err(e) = self
-            .watcher
-            .watch(self.path.as_ref(), notify::RecursiveMode::NonRecursive)
-        {
-            error!("Failed to set watch: {}", e);
-        } else {
-            trace!("Set watch");
+enum Empty {}
+
+async fn watcher(w: PolicySetWatcher) -> Result<Empty> {
+    let mut last_modified = get_last_modified(&w.policy_set).await?;
+    loop {
+        tokio::time::sleep(Duration::from_secs(1)).await;
+        let time = get_last_modified(&w.policy_set).await?;
+        if time != last_modified {
+            match attempt_policy_reload(&w).await {
+                Ok(policies) => {
+                    send_query(policies, &w.tx).await?;
+                }
+                Err(e) => error!("Error reloading policies: {e}"),
+            };
+            last_modified = time;
```

Comment:
Don't update the time unless it succeeds? Will presumably be Ok as is, but this looks funny to me.